content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning.md

```diff
@@ -31,9 +31,65 @@ level of your CustomResourceDefinitions or advance your API to a new version wit
+to new versions of a CustomResourceDefinition.
+
+When a CustomResourceDefinition is first created, the first version is set at
+an appropriate stability level and a version number. For example `v1beta1`
+would indicate that the first version is not yet stable. All custom resource objects
+will initially be stored at this version.
+
+Once the CustomResourceDefinition is created, clients may begin using the
+`v1beta1` API.
+
+Later it might be necessary to add new version such as `v1`.
+
+Adding a new version:
+1. Select a conversion strategy. Since custom resource objects need to be able
+   to be served at both versions, that they will sometimes be served at a
+   different version they are stored at. In order for this to be possible, the
+   custom resource objects must sometimes be converted between the version they
+   are stored at and the version they are served at. If the conversion involves
+   schema changes and requires custom logic, a conversion webhook should be
+   used. If there are no schema changes, the default `None` conversion strategy
+   may be used.
+2. If using conversion webhooks, create and deploy the conversion webhook. See
+   the [Webhook conversion](#webhook-conversion) for more details.
+3. Update the CustomResourceDefinition to serve both the old and new versions
+   and to use the selected conversion strategy. If using a conversion webhook,
+   also set the webhook configuration.
+
+Once the new version is added, clients may incrementally migrated to the `v1`
+new version. It is perfectly safe for some clients to use `v1beta1` while others
+use `v1`. It is safe for clients to use both versions before, during and after
+the storage migration.
+
+Storage Migration:
+1. Update the version the custom resources objects are stored at in the
+   CustomResourceDefinition to the new version. Changes to the storage version
+   take effect for each custom resource instance the next time that resource is
+   written.
+1. Complete the storage migration
+  - Option 1: Use [Storage Version Migrator](https://github.com/kubernetes-sigs/kube-storage-version-migrator)
+  - Option 2: Write to each custom resource object at least once. A get followed by a put of the identical
+    content is sufficient. No modification of the data is required.
+
+Removing an old version:
+1. Ensure all clients are fully migrated to the new version. The kube-apiserver
+   logs can reviewed to help identify any clients that are still accessing via the
+   old version.
+1. Ensure CustomResourceDefinition is configured to store custom resource objects at the new version
+1. Ensure the storage migration, described above, has been completed. If using a
+   conversion webhook, check if it is recieving any conversions requests when
+   making list requests for custom resources at the old version. If not, all
+   custom resource objects have been converted.
+1. Remove the old version from the CustomResourceDefinition
```

Comment:
be specific here (e.g. "... from the CustomResourceDefinition `spec.versions` list")